### PR TITLE
Read only property

### PR DESCRIPTION
**Pull request by gf712**
@karlnapf Is this what you had in mind?

### CODE REVIEW

**Comment by gf712** [comment]
We can't call the enum field `CONST` because msvc thinks we are using the keyword `const`...

**Comment by gf712** [comment]
Should a constant parameter be `put`-able once for initialisation, or would it always be read-only?

File: src/shogun/base/AnyParameter.h [hunk 1]
```diff
@@ -37,7 +37,8 @@ namespace shogun
 		HYPER = 1u << 0,
 		GRADIENT = 1u << 1,
 		MODEL = 1u << 2,
-		AUTO = 1u << 10
+		AUTO = 1u << 10,
+		CONST = 1u << 11
```
Comment (lisitsyn) [inline src/shogun/base/AnyParameter.h]: `READ_ONLY` might be a better name to avoid clashes

**Comment by karlnapf** [comment]
I think always read only, never put able. And then we try to use factories.... let’s see if that approach works at least. Can you mark feature matrix of dense feats also read only?

File: src/shogun/base/AnyParameter.h (hunk 1 above)
Comment (gf712) [reply src/shogun/base/AnyParameter.h]: yup, I didn't realise that it could clash with C++ keywords...

**Comment by gf712** [comment]
So should the kernel ,for example, only be initialised with `init(lhs,rhs)`, instead of `sg.kernel("name", lhs=lhs, rhs=rhs)`?

**Comment by karlnapf** [comment]
Exactly! That’s the the point

File: examples/meta/src/gaussian_process/regression.sg [hunk 2]
```diff
@@ -11,7 +11,8 @@ Labels labels_test = labels(f_labels_test)
 #![create_features]
 
 #![create_appropriate_kernel_and_mean_function]
-Kernel k = kernel("GaussianKernel", lhs=features_train, rhs=features_train, log_width=0.0)
+Kernel k = kernel("GaussianKernel", log_width=0.0)
```
Comment (karlnapf) [inline examples/meta/src/gaussian_process/regression.sg]: yes!

File: src/shogun/kernel/Kernel.cpp [hunk 3]
```diff
@@ -922,7 +922,8 @@ void CKernel::register_params()   {
 	    "Cache size in MB.");
 
 	SG_ADD(
-	    &lhs, "lhs", "Feature vectors to occur on left hand side.");
+		&lhs, "lhs", "Feature vectors to occur on left hand side.",
+		ParameterProperties::READONLY);
```
Comment (karlnapf) [inline src/shogun/kernel/Kernel.cpp]: rhs should also be read-only

**Comment by vigsterkr** [comment]
@karlnapf i'm just wondering in case of DenseFeatures the m_feature could be defined as a `const` and that should have the same effect... or?

**Comment by karlnapf** [comment]
Unfortunately no. It was my preferred solution as well but the any framework makes the compiler freak out if we do that (e.g clone, put) and then when I talked to @lisitsyn and he had some other reasons why not to do that. This pr here is the result of our discussion

**Comment by lisitsyn** [comment]
@vigsterkr I am just afraid adding constness would make things more complex. It could make sense in pure C++ but since we pass the runtime boundary in the interfaces it might be not that good.

**Comment by lisitsyn** [comment]
Ah and this also nearly doubles the amount of generated template code.

File: src/shogun/kernel/Kernel.cpp [hunk 4]
```diff
@@ -920,11 +920,14 @@ void CKernel::save_serializable_post() throw (ShogunException)
...
-	SG_ADD(&rhs, "rhs", "Feature vectors to occur on right hand side.");
 	SG_ADD(
-	    &lhs_equals_rhs, "lhs_equals_rhs",
-	    "If features on lhs are the same as on rhs.");
+		&lhs, "lhs", "Feature vectors to occur on left hand side.",
+		ParameterProperties::READONLY);
+	SG_ADD(
+		&rhs, "rhs", "Feature vectors to occur on right hand side.",
+		ParameterProperties::READONLY);
+	SG_ADD(&lhs_equals_rhs, "lhs_equals_rhs",
```
Comment (karlnapf) [inline src/shogun/kernel/Kernel.cpp]: also readonly!

**Comment by karlnapf** [comment]
CI problems?

**Pull request merged**